core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/io/SdkLengthAwareInputStream.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core.internal.io;
+
+import static software.amazon.awssdk.utils.NumericUtils.saturatedCast;
+
+import java.io.FilterInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * An {@code InputStream} that is aware of its length. The main purpose of this class is to support truncating streams to a
+ * length that is shorter than the total length of the stream.
+ */
+@SdkInternalApi
+public class SdkLengthAwareInputStream extends FilterInputStream {
+    private long length;
+    private long remaining;
+
+    public SdkLengthAwareInputStream(InputStream in, long length) {
+        super(in);
+        this.length = Validate.isNotNegative(length, "length");
+        this.remaining = this.length;
+    }
+
+    @Override
+    public int read() throws IOException {
+        if (!hasMoreBytes()) {
+            return -1;
+        }
```

Comment:
Would it make sense to add a debug statement here?

Reply:
Sure makes sense